R/deps_installation_proposal.R

```diff
@@ -299,6 +302,56 @@ get_refs_from_desc <- function(d) {
   res[res_idx]
 }
 
+#' Replace Remotes in the `desc` that have been resolved to a GitHub tag or are
+#' in CRAN
+#'
+#' Replaces any existing Remotes entry with the resolved GitHub tag from the
+#' `new_refs`.
+#'
+#' It keeps all the existing Remotes that have not been resolved in `new_refs`.
+#'
+#' @param d (`desc`) DESCRIPTION object
+#' @param new_refs (`list`) remote references that have been resolved and are
+#' being updated in `Config/Needs/verdepcheck`
+#' @keywords internal
+desc_remotes_cleanup <- function(d, new_refs) {
+  # Parse the remotes to retrieve the package names
+  remotes <- pkgdepends::parse_pkg_refs(d$get_remotes())
+
+  # Get the packages defined in remotes
+  #  (making sure that only packages that are already defined here are modified)
+  remotes_pkg <- vapply(remotes, `[[`, character(1), "package")
+
+  # Find which packages of the new_refs are defined in Remotes
+  new_refs_gh <- vapply(
+    new_refs,
+    function(.x) {
+      isTRUE(.x$package %in% remotes_pkg) && inherits(.x, "remote_ref_github")
+    },
+    logical(1)
+  )
+
+  # New remotes ref to use when replacing
+  new_ref_remote <- vapply(new_refs[new_refs_gh], `[[`, character(1), "ref")
+
+  # Remotes that are not in new_refs are kept, as well as the ones that were
+  #  resolved to be a github repo
+  d$clear_remotes()
+
+  new_ref_pkg <- vapply(new_refs, `[[`, character(1), "package")
+  new_remotes <- c(
+    # Keep remotes
+    d$get_remotes()[!(remotes_pkg %in% new_ref_pkg)],
+    # Modified remotes
+    new_ref_remote
+  )
```

Comment:
generally speaking it seems that it's just `new_remotes_refs <- setdiff(remotes_refs, new_refs)` - i.e. remove remotes that exists in the second fun arg. You seem to do `new_remotes_refs <- c(setdiff(remotes_refs, setdiff(new_refs, remotes_refs)), setdiff(new_refs, remotes_refs))` which is a little bit more complex to follow and fully understand. Am I right with my thinking? Is there any reason why we are doing it this way?

Reply:
The idea here is to remove all packages that have been resolved to a CRAN package and modify the ones that resolved to a GitHub tag.

> `new_remotes_refs <- c(setdiff(remotes_refs, setdiff(new_refs, remotes_refs)), setdiff(new_refs, remotes_refs))`

I think the code below better shows the logic _(although `setdiff` cannot be used as we're comparing package name, but use remote reference)_

note: the `setdiff` should only output a non-empty vector if the DESCRIPTION is badly formed

```R
c(
  setdiff(remotes_refs, new_refs), # should always be empty unless DESCRIPTION is badly formed
  intercept(remote_refs, new_refs[is.github.tag])
)
```

Transforming:

```R
# (...)
Imports:
  a,
  b,
  c
Remotes:
  ie/b@*release
  ie/c@*release
Config/Need/verdepcheck:
  ie/a,
  ie/b,
  ie/c
```

To:

```R
# (...)
Imports:
  a (>= 1.0.0),
  b (>= 2.0.0),
  c (>= 3.0.0)
Remotes:
  b=ie/b@v2.0.0
Config/Need/verdepcheck:
  a,
  b=ie/b,
  c # exists on CRAN, this is done by `desc_cond_set_refs()`
```